### PR TITLE
Feature - Firebase Analytics & Crashlytics

### LINKED ISSUE #15: Need to add tool for Crash Reporting & Analytics
I'm familiar with Sentry and it works well for Flutter in prev project so I can set up Sentry account and add to app unless someone else has different preferences?
cc: @workerbee22 @alessandrofavero @adamkoch @domesticmouse

### PR DESCRIPTION
**Pull request by workerbee22**
Initial implementation of Firebase Analytics (data is also passed to Google Analytics) and Firebase Crashlytics for crash & error reporting. Note this is just the initial implementation.

## Changes
- Firebase Analytics setup for iOS and Android
- Firebase Crashlytics setup for iOS and Android for uncaught, Flutter and Crashlytics trapped errors
- Analytics flagging each screen in the app as a screen, using constants and separate to navigation observer to ensure maintainable data in analytics as the app changes.

## Screenshots
No UI changes.

## Things to note
- Release notes not updated as no user facing changes.
- Analytics event method is present but not yet applied to event like playing PPE videos
- Firebase console and Google Analytics console setup also accompany this PR.
- Removed an orphan/duplicate file

### CODE REVIEW

File: lib/constants.dart
Comment (maks) [inline lib/constants.dart:11]: can these just be the route name consts to keep things more DRY ?

File: lib/utils/firebase.dart
Comment (maks) [inline lib/utils/firebase.dart]: could we at least just move this into a CrashSimlutator class please? if we want to keep code then lets make sure its not commented out so all the usual tooling works with it.

File: lib/utils/firebase.dart
Comment (maks) [inline lib/utils/firebase.dart]: Its a bit strange to have a separate init method thats only called from the constructor, is there a reason for that?

File: lib/view/airway/checklist/intubation_checklist_infographic_page.dart
Comment (maks) [inline lib/view/airway/checklist/intubation_checklist_infographic_page.dart:11]: I'm not very familiar with Google analytics client, but if its already tracking route navigation events, what does this add as there are no other analytics events being tracked on this screen?

File: lib/view/airway/checklist/intubation_checklist_page.dart
Comment (maks) [inline lib/view/airway/checklist/intubation_checklist_page.dart]: ditto above for all the other calls to Analytics().analyticsScreen()

File: lib/main.dart
Comment (maks) [inline lib/main.dart]: it looks quite strange just calling a constructor and then not doing anything with the object itself. Since CrashReporting has no instance properties can it just be a static function?

File: ios/Runner.xcodeproj/project.pbxproj
Comment (sallychung) [inline ios/Runner.xcodeproj/project.pbxproj]: Purely for project config cleanliness.
The comment in the script doesn’t make sense in here as it’s run in the build phase. Can we please remove it? 
Doesn’t have to be done in this PR

File: lib/view/icu_non_intensivist/daily_round/daily_round_view.dart
Comment (maks) [inline lib/view/icu_non_intensivist/daily_round/daily_round_view.dart:13]: :+1:

**Review by maks — CHANGES_REQUESTED** [review]

File: lib/constants.dart
Comment (workerbee22) [reply lib/constants.dart:11]: @maks This was very much a conscious decision, because one of the poor people who have to review analytics, routes change a lot as the app changes, name changes, re routes, changed routes etc.Which means over a short time the route based method breaks down and it becomes a mess of new, newer, newest, older etc.

So this way the analytics screens should mostly stay fixed and independent as the routes all inevitably change.

File: lib/utils/firebase.dart
Comment (workerbee22) [reply lib/utils/firebase.dart]: @maks So you'd prefer separation of initialisation and simulation?

File: lib/utils/firebase.dart
Comment (workerbee22) [reply lib/utils/firebase.dart]: @maks Not really ... what would you prefer?

File: lib/view/airway/checklist/intubation_checklist_infographic_page.dart
Comment (workerbee22) [reply lib/view/airway/checklist/intubation_checklist_infographic_page.dart:11]: See above ☝️

File: lib/view/airway/checklist/intubation_checklist_page.dart
Comment (workerbee22) [reply lib/view/airway/checklist/intubation_checklist_page.dart]: See above ☝️

File: lib/main.dart
Comment (workerbee22) [reply lib/main.dart]: I'll check

File: lib/view/icu_non_intensivist/daily_round/daily_round_view.dart
Comment (workerbee22) [reply lib/view/icu_non_intensivist/daily_round/daily_round_view.dart:13]: Yep a reversion.

**Comment by lukesleeman** [comment]
I've tested the branch out, and it appears to have no ill effects.  Kind of hard to tell if its doing anything or not 😁

**Comment by workerbee22** [comment]
@lukesleeman Yep, until there is some data, it's a nice to have thing. 

But using the simulation code (in there but commented out) one crash type at a time and setting `Crashlytics.instance.enableInDevMode = true` you can see the crashes locally on your console. 

So I've tested all 3 levels of crash locally on Android and iOS for the simulations. Just waiting on crash data to show up in the console ... but its been a while so will go back and check. But if the app isn't talking to Crashlytics then when you goto the Firebase console you only get the setup wizard'. So because we can see the proper Crashlytics console view the app MUST have talked to Crashlytics already.

**Review by sallychung — APPROVED** [review]
Seems to be working. Tested through Proxyman

image

File: lib/constants.dart
Comment (maks) [reply lib/constants.dart:11]: Hmm, ok, I've never been a consumer of analytics so will defer to those that do :+1:

File: lib/main.dart
Comment (maks) [reply lib/main.dart]: So consolidating the comment about the constructor with this one: can really just be a static function, no class needed

File: lib/utils/firebase.dart
Comment (maks) [reply lib/utils/firebase.dart]: yes the simulation is only for local debugging and testing so if its useful to keep once this is ready for merging then yes please the simulation should be separate from the production code that does the initialisation

File: ios/Runner.xcodeproj/project.pbxproj
Comment (yimingc2) [inline ios/Runner.xcodeproj/project.pbxproj:213]: Is the script for Xcode 10? After 30th April, apps that submitted to App Store must use iOS 13 SDK or later which is Xcode 11+.

**Comment by maks** [comment]
>So I've tested all 3 levels of crash locally on Android and iOS for the simulations. Just waiting on crash data to show up in the console ... but its been a while so will go back and check. But if the app isn't talking to Crashlytics then when you goto the Firebase console you only get the setup wizard'. So because we can see the proper Crashlytics console view the app MUST have talked to Crashlytics already.

@workerbee22  So I'd feel *much* better once we can actual app generated errors, especially app crashes in the console :slightly_smiling_face:  
as this is *exactly* what I experienced previously trying and failing to use Crashlytics in a Flutter app: everything looked fine locally but nothing ever showed up in the web console...

**Comment by workerbee22** [comment]
I'll do a check against the project for which I provided the screen shots ie. these. Perhaps I missed something:

Issue: https://github.com/Western-Health-Covid19-Collaboration/wh_covid19_app/issues/15

image

**Review by yimingc2 — APPROVED** [review]

File: lib/utils/firebase.dart
Comment (yimingc2) [inline lib/utils/firebase.dart]: We can make it `static` and then we don't need to create a new instance of this class every time.

**Comment by yimingc2** [comment]
Looks good to me! Appreciate for your effort to make it happen!

File: ios/Runner.xcodeproj/project.pbxproj
Comment (workerbee22) [reply ios/Runner.xcodeproj/project.pbxproj]: Yep ... it was the default. I'll remove 👍

File: ios/Runner.xcodeproj/project.pbxproj
Comment (workerbee22) [reply ios/Runner.xcodeproj/project.pbxproj:213]: That was auto generated, so no idea.

File: lib/utils/firebase.dart
Comment (workerbee22) [reply lib/utils/firebase.dart]: oh yeah should be ... oops 👍@maks flagged it too.

**Comment by workerbee22** [comment]
OK so nothing in the console yet for Crashlytics from my tests ... investigating.

Analytics data coming through fine.

File: lib/utils/firebase.dart
Comment (workerbee22) [reply lib/utils/firebase.dart]: Ok abstracted into `CrashSimulation` class.

File: lib/main.dart
Comment (workerbee22) [reply lib/main.dart]: ok just a static function now.

File: ios/Runner.xcodeproj/project.pbxproj
Comment (workerbee22) [reply ios/Runner.xcodeproj/project.pbxproj]: Removed comment.

File: lib/utils/firebase.dart
Comment (workerbee22) [reply lib/utils/firebase.dart]: done

**Comment by workerbee22** [comment]
OK PR workflow build all OK. Merging this one.

**Pull request merged**
